tools/pubsys/src/aws/client.rs

```diff
@@ -1,159 +1,169 @@
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+use aws_config::{ConfigLoader, SdkConfig};
+use aws_sdk_ec2::{Endpoint, Region};
+use aws_sdk_sts::model::Credentials as StsCredentials;
+use aws_sdk_sts::{Client as StsClient, Credentials as SdkCredentials};
+use pubsys_config::AwsConfig as PubsysAwsConfig;
+use snafu::{OptionExt, ResultExt};
+use std::convert::TryInto;
 
-impl NewWith for Ec2Client {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+const PROVIDER_NAME: &str = "pubsys";
 
-impl NewWith for SsmClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
-
-impl NewWith for StsClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
-
-/// Create a rusoto client of the given type using the given region and configuration.
-pub(crate) fn build_client<T: NewWith>(
+pub(crate) async fn build_client_config(
     region: &Region,
     sts_region: &Region,
-    aws: &AwsConfig,
-) -> Result<T> {
-    let maybe_regional_role = aws.region.get(region.name()).and_then(|r| r.role.clone());
-    let assume_roles = aws.role.iter().chain(maybe_regional_role.iter()).cloned();
-    let provider = build_provider(
-        sts_region,
-        assume_roles.clone(),
-        base_provider(&aws.profile)?,
-    )?;
-    Ok(T::new_with(
-        rusoto_core::HttpClient::new().context(error::HttpClientSnafu)?,
-        provider,
-        region.clone(),
-    ))
-}
+    pubsys_aws_config: &PubsysAwsConfig,
+) -> Result<SdkConfig> {
+    let mut config = aws_config::from_env();
+
+    // Add custom endpoint if specified in aws.region.REGION.endpoint.
+    let maybe_regional_endpoint = pubsys_aws_config
+        .region
+        .get(region.as_ref())
+        .and_then(|r| r.endpoint.clone());
+    if let Some(endpoint) = &maybe_regional_endpoint {
+        config = config.endpoint_resolver(Endpoint::immutable(
+            endpoint
+                .parse()
+                .context(error::UriParseSnafu { input: endpoint })?,
+        ));
+    }
```

Comment:
The SDK comment for `endpoint_resolver` says:
```
/// This method will override the endpoint resolver used for **all** AWS services. This mainly                                               
/// exists to set a static endpoint for tools like `LocalStack`. For live traffic, AWS services                                              
/// require the service-specific endpoint resolver they load by default.    
```

Do we actually use that endpoint functionality for anything today? If not, we should just retract the option in `PubsysAwsConfig`.

To handle it correctly, we'd need service-specific endpoint fields in `Infra.toml`, and SDK clients would need to be built to use those endpoints. That would happen in the calling code rather than here.

Reply:
I looked into it, but I wasn't able to find anywhere we are using it today.